.github/workflows/run-tests.yml

```
@@ -50,6 +63,17 @@ jobs:
 
       - run: make build-tests
 
+      - name: build functional test matrix
+        id: json
+        run: |
+          if [ ${{ inputs.run_single_functional_test }} = 'true' ]; then shard_index='{ "shard_index": [0] }'; else shard_index='{ "shard_index": [0, 1, 2] }'; fi
+          echo "shard_index=$shard_index" >> "$GITHUB_OUTPUT"
+          if [ ${{ inputs.run_single_functional_test }} = 'true' ]; then total_shards=1; else total_shards=3; fi
+          echo "total_shards=$total_shards" >> "$GITHUB_OUTPUT"
+          if [ ${{ inputs.run_single_functional_test }} = 'true' ]; then single_test_args="-test.run ${{ inputs.test_name }} -count ${{ inputs.n_runs }}"; fi
+          echo "single_test_args=$single_test_args" >> "$GITHUB_OUTPUT"
```

just to make this easier to read, how about:
```suggestion
          shards=3
          if [[ ${{ inputs.run_single_functional_test }} = true ]]; then
            shards=1
            single_test_args="-test.run ${{ inputs.test_name }} -count ${{ inputs.n_runs }}"
          fi
          {
            echo "shard_index={\"shard_index\":[$(seq -s, 0 $((shards-1)))]}"
            echo "total_shards=$shards"
            echo "single_test_args=$single_test_args" 
          } >> "$GITHUB_OUTPUT"
```

---

I changed some other stuff so it's different now since I split this up into different steps. I might merge into one step though so that I only have to do one if check as you suggest here.